### PR TITLE
JS: Don't flag optionally sanitized XSS sinks.

### PR DESCRIPTION
**Pull request by erik-krogh**
Consider the below example.  

```JavaScript
var source = tainted();
if (sanitize) {
  source = DOMPurify.sanitize(source);
}
$("#something").html(source);
```

`js/xss`/`js/xss-through-dom` flags this because `search` is only optionally sanitized.  
With this PR the data-flow edge for `search` from the first to the last line is removed, and thus `js/xss` no longer flags this example.  

Data-flow is still allowed through a bad sanitizer. 

With this PR we get a TN for CVE-2018-19048 and CVE-2019-8331. 

An edge like the one in the above example [appears somewhat often](https://lgtm.com/query/1138415514053724668/) (although most of those are copies of `angular.js`). 

[A small evaluation](https://git.semmle.com/erik/dist-compare-reports/tree/profiling-erik-krogh.northeurope.cloudapp.azure.com_1590415691999) looks fine performance wise.    
[An evaluation on nightly](https://git.semmle.com/erik/dist-compare-reports/tree/profiling-erik-krogh.northeurope.cloudapp.azure.com_1590463903416) also looks fine. (and [a redo of the outliers](https://git.semmle.com/erik/dist-compare-reports/tree/profiling-erik-krogh.northeurope.cloudapp.azure.com_1590480763189) shows that the regressions were just performance wobble). 

I also added a sanitizer-guard for `x.length === 0` to recognize [this sanitizer](https://github.com/twbs/bootstrap/blob/7bc4d2e0bc65151b6f60dccad50c9c8f50252bd6/js/src/tools/sanitizer.js#L93). 

I only mentioned `js/xss` in the change note, as I don't think we need to repeat the same thing for `js/xss-through-dom`.

### CODE REVIEW

File: javascript/ql/src/semmle/javascript/security/dataflow/Xss.qll [hunk 1]
```diff
@@ -302,6 +302,38 @@ module DomBasedXss {
   private class MetacharEscapeSanitizer extends Sanitizer, Shared::MetacharEscapeSanitizer { }
 
   private class UriEncodingSanitizer extends Sanitizer, Shared::UriEncodingSanitizer { }
+
+  /**
+   * Holds if there exists two dataflow edges to `succ`, where one edges is sanitized, and the other edge starts with `pred`. 
+   */
+  predicate isOptionallySanitizedEdge(DataFlow::Node pred, DataFlow::Node succ) {
+    exists(DataFlow::CallNode sanitizer |
+      sanitizer.getCalleeName().regexpMatch("(?i).*sanitize.*")
```
Comment (asgerf) [inline javascript/ql/src/semmle/javascript/security/dataflow/Xss.qll:311]: Could you use `HtmlSanitizerCall` instead, and extend it to ensure `DOMPurify.sanitize` is matched?

Comment (erik-krogh) [reply javascript/ql/src/semmle/javascript/security/dataflow/Xss.qll:311]: One of the CVEs use their own sanitizer, so we would not recognize that if we use `HtmlSanitizerCall`.  

How about I add `or sanitizer instanceof HtmlSanitizerCall`?

Comment (asgerf) [reply javascript/ql/src/semmle/javascript/security/dataflow/Xss.qll:311]: Would the name heuristic in `HtmlSanitzerCall` match that?

Comment (erik-krogh) [reply javascript/ql/src/semmle/javascript/security/dataflow/Xss.qll]: > Would the name heuristic in `HtmlSanitzerCall` match that?

It does, I missed that when I looked at `HtmlSanitzerCall` the first time.  
I implemented your suggestion. We still get TN on the 2 CVEs.

File: javascript/ql/src/semmle/javascript/HtmlSanitizers.qll [hunk 2]
```diff
@@ -48,6 +48,11 @@ private class DefaultHtmlSanitizerCall extends HtmlSanitizerCall {
       or
       callee = LodashUnderscore::member("escape")
       or
+      exists(DataFlow::PropRead read | read = callee |
+        read.getPropertyName() = "sanitize" and
+        read.getBase().asExpr().(VarAccess).getName() = "DOMPurify"
```
Comment (asgerf) [inline javascript/ql/src/semmle/javascript/HtmlSanitizers.qll:53]: Could we base this on `moduleImport` and `globalVarRef` instead?

Comment (erik-krogh) [reply javascript/ql/src/semmle/javascript/HtmlSanitizers.qll:53]: That could work, but only if we type-tracked the `moduleImport`.   
(This is where DOMPurify is required: https://github.com/mycolorway/simditor/blob/ef01a643cbb7f8163535d6bfb71135f80ec6a6fd/lib/simditor.js#L16). 

So for now, I think I prefer to keep it as is.

Comment (asgerf) [reply javascript/ql/src/semmle/javascript/HtmlSanitizers.qll:53]: Good point. I've added a note about it in https://github.com/github/codeql-javascript-team/issues/118#issuecomment-633989671

File: change-notes/1.25/analysis-javascript.md [hunk 3]
```diff
@@ -36,6 +36,7 @@
 
 | **Query**                      | **Expected impact**          | **Change**                                                                |
 |--------------------------------|------------------------------|---------------------------------------------------------------------------|
+| Client-side cross-site scripting (`js/xss`) | Less results | This query no longer flags optionally sanitized values. |
```
Comment (asgerf) [inline change-notes/1.25/analysis-javascript.md:39]:
```suggestion
| Client-side cross-site scripting (`js/xss`) | Fewer results | This query no longer flags optionally sanitized values. |
```

**Pull request merged**